### PR TITLE
API: Raise when setting name via level

### LINKED ISSUE #29032: Deprecate getting the name from MultiIndex.levels
Follow up to https://github.com/pandas-dev/pandas/pull/27242.

https://github.com/pandas-dev/pandas/pull/27242#issuecomment-542602004

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes https://github.com/pandas-dev/pandas/issues/29032

This is extremely ugly, but gets the job done. It's probably worth doing to avoid silently failing, but not sure.

cc @topper-123.

### CODE REVIEW

File: pandas/core/indexes/interval.py [hunk 1]
```diff
@@ -234,6 +234,7 @@ def _simple_new(cls, array, name, closed=None):
         result = IntervalMixin.__new__(cls)
         result._data = array
         result.name = name
+        result._no_setting_name = False
```
Comment (jbrockmendel) [inline pandas/core/indexes/interval.py]: does this need to be set to False in each _simple_new, or can it just use the class-level default?

File: pandas/tests/indexes/multi/test_names.py [hunk 2]
```diff
@@ -124,3 +124,12 @@ def test_get_names_from_levels():
 
     assert idx.levels[0].name == "a"
     assert idx.levels[1].name == "b"
+
+
+def test_setting_names_from_levels_raises():
+    idx = pd.MultiIndex.from_product([["a"], [1, 2]], names=["a", "b"])
+    with pytest.raises(RuntimeError, match="set_names"):
+        idx.levels[0].name = "foo"
```
Comment (jbrockmendel) [inline pandas/tests/indexes/multi/test_names.py]: i could do something weird like:

```
lvl = idx.levels[0]

ser = pd.Series(lvl, index=lvl)
ser.index.name="foo"
```

is there a less-contrived way in which `idx.levels[0]` could come un-attached from the multiindex?

File: pandas/core/indexes/base.py [hunk 3]
```diff
@@ -1214,6 +1219,12 @@ def name(self):
 
     @name.setter
     def name(self, value):
+        if self._no_setting_name:
+            # Used in MultiIndex.levels.
+            raise RuntimeError(
```
Comment (topper-123) [inline pandas/core/indexes/base.py]: TypeError instead?

**Review by topper-123 — COMMENTED** [review]
If you can avoid setting this in subclasses (other than MultiIndex), it would simplify quite a bit.

This will be removed later, when we feel confident that people have switched idiom, right? It's in general not possible to prevent people setting attributes on child objects (e.g. in CategoricalIndex, people can already do ``cidx.categories.name = "name"``) and it just adds complexity trying to prevent it, IMO.

**Comment by TomAugspurger** [comment]
I think it has to be on all index subclasses since they make up the levels of a MultiIndex.

File: pandas/core/indexes/interval.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/core/indexes/interval.py]: If it’s just on the class, won’t it be shared amongst all instances? I think I should change the type declaration to not have a default.

**Comment by topper-123** [comment]
@TomAugspurger , but if it's set on ``Index`` all index subclasses inherit the attribute?

File: pandas/core/indexes/interval.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/interval.py]: > If it’s just on the class, won’t it be shared amongst all instances?

I think the relevant thing is: when you set `level._no_setting_name = True` in the one place where you do that, will that update the class attribute, and the answer should be no.

Comment (topper-123) [reply pandas/core/indexes/interval.py]: Yeah, ``level._no_setting_name = True`` adds an instance attribute, so there shouldn't happen anything on the class attribute.

**Comment by TomAugspurger** [comment]
Right, since it’s the level of the MI that can’t have its name set, not the MI itself.

File: pandas/tests/indexes/multi/test_names.py (hunk 2 above)
Comment (TomAugspurger) [reply pandas/tests/indexes/multi/test_names.py]: That raises (correctly, I think). I can add a test for it if you like.

File: pandas/core/indexes/base.py (hunk 3 above)
Comment (TomAugspurger) [reply pandas/core/indexes/base.py]: I think a RuntimeError is best, since we don't check the type or valuje of `value` checked here.

**Comment by TomAugspurger** [comment]
My apologies, I've simplified things a bit. I misunderstood what happened to class attributes when the attribute is set on an instance.

**Comment by TomAugspurger** [comment]
> This will be removed later, when we feel confident that people have switched idiom, right?

I'm not sure. We can weigh the maintenance cost down the road and remove it if it's burdensome.

**Comment by jbrockmendel** [comment]
I'm fine with this as-is.

Spitballing another solution: we could have Index._names be an ndarray for all Index subclasses, with

```
@property
def name(self):
    return self._names[0]     

@name.setter
def name(self, value):
    if np.is_view_or_whatever(self._names):
        raise RuntimeError(...)
    self._names[0] = value
```

**Comment by jreback** [comment]
this looks fine, butI would consider something slightly different. We could have an attribute defined on Index._parent = None

then set a weakref when it is contained in a MultiIndex. This gives us provenance on where things are coming from. might be useful.

**Comment by TomAugspurger** [comment]
Mmm, that's also reasonable. My initial concern is that indexes are potentially shared, so they could have multiple parents? For this purpose it's not a problem since we're creating the Index from scratch, just something to keep in mind going forward.

**Comment by jreback** [comment]
> Mmm, that's also reasonable. My initial concern is that indexes are potentially shared, so they could have multiple parents? For this purpose it's not a problem since we're creating the Index from scratch, just something to keep in mind going forward.

I don’t think we share for a contained Index inside a MultiIndex but not really sure

**Comment by TomAugspurger** [comment]
You're right, these levels of a MI are created on the fly.

In theory, a user-action could end up with an Index that's created from the level of a MI

```python
s = pd.Series(index=idx.levels[0])
```

but that probably doesn't affect your proposed "parent" reference.

That implementation is going to be more complex though, since subclasses don't call `Index._simple_new`. Thoughts on waiting till we need it?

**Comment by jreback** [comment]
yeah was just an idea i think what u did here is ok
let me have another glance

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @TomAugspurger

**Pull request merged**